src/parallel.jl

```diff
@@ -0,0 +1,228 @@
+Julia with multiple threads with the `-t n` option. If your data
+loading is bottlenecked by the CPU, it is recommended to set `n`
+to the number of physical CPU cores.
+
+## Arguments
+
+- `data`: a data container that implements `getindex/getobs` and `length/numobs`
+- `buffer = false`: whether to use inplace data loading with `getobs!`. Only use
+    this if you need the additional performance and `getobs!` is implemented for
+    `data`. Setting `buffer = true` means that when using the iterator, an
+    observation is only valid for the current loop iteration.
+- `executor = Folds.ThreadedEx()`: task scheduler
+    You may specify a different task scheduler which can
+    be any `Folds.Executor`.
+- `channelsize = Threads.nthreads()`: the number of observations that are prefetched.
+    Increasing `channelsize` can lead to speedups when per-observation processing
+    time is irregular but will cause higher memory usage.
+"""
+function eachobsparallel(
+        data,
+        executor::Executor = _default_executor();
+        buffer = false,
+        channelsize = Threads.nthreads())
+    if buffer
+        return _eachobsparallel_buffered(data, executor; channelsize)
+    else
+        return _eachobsparallel_unbuffered(data, executor; channelsize)
+    end
+end
+
+
+function _eachobsparallel_buffered(data, executor; channelsize=Threads.nthreads())
+    # Prepare initial buffers
+    buf = getobs(data, 1)
+    buffers = [buf]
+    foreach(_ -> push!(buffers, deepcopy(buf)), 1:channelsize)
+
+    # This ensures the `Loader` will take from the `RingBuffer`s result
+    # channel, and that a new results channel is created on repeated
+    # iteration. (Since `Loader`) closes the previous at the end of
+    # each iteration.
+    setup_channel(sz) = RingBuffer(buffers)
+
+    return Loader(1:numobs(data); executor, channelsize, setup_channel) do ringbuffer, i
+        # Internally, `RingBuffer` will `put!` the result in the results channel
+        put!(ringbuffer) do buf
+            getobs!(buf, data, i)
+        end
+    end
+end
+
+function _eachobsparallel_unbuffered(data, executor; channelsize=Threads.nthreads())
+    return Loader(1:numobs(data); executor, channelsize) do ch, i
+        obs = getobs(data, i)
+        put!(ch, obs)
+    end
+end
+
+
+# Unlike DataLoaders.jl, this currently does not use task pools
+# since  `ThreadedEx` has shown to be more performant. This may
+# change in the future.
+# See PR 33 https://github.com/JuliaML/MLUtils.jl/pull/33
+_default_executor() = ThreadedEx(basesize=1)
+
+
+# ## Internals
+
+# The `Loader` handles the asynchronous iteration and fills
+# a result channel.
+
+
+"""
+    Loader(f, args; executor, channelsize, setup_channel)
+
+Create a threaded iterator that iterates over `(f(arg) for arg in args)`
+using threads that prefill a channel of length `channelsize`.
+
+Note: results may not be returned in the correct order, depending on
+`executor`.
+"""
+mutable struct Loader
```

Comment:
Doesn't look like this needs to be mutable? Also, is there no value in concretely typing any of the fields?

Reply:
True, the `mutable` is no longer necessary. I'll add types for `executor`, restrict `argiter` to an `AbstractVector` and channelsize to `Int` when I'm back at my PC